### ISSUE TITLE
[BUG] HasMany relationship not stored in database when two relationship fields are on the same table

### ISSUE THREAD

**Issue by madriegh** [Laravel-Backpack/CRUD#4552]
**(A) what you did**
I have a table with transactions (model FK, label, amount, type) in my database. They are linked to a model and can be of type 'income' or 'expense'.
I have the following relations on this model.
```
    public function incomes()
    {
        return $this->hasMany(Transaction::class)->ofType('income');
    }

    public function expenses()
    {
        return $this->hasMany(Transaction::class)->ofType('expense');
    }
```
where ofType is a query scope.
In the ModelCrudController I have one relationship field for the incomes and one for the expenses.
```
        CRUD::addField([
            'type'          => 'relationship',
            'name'          => 'incomes',
            'subfields'   => [
                [
                    'name' => 'label',
                    'type' => 'text',
                ],
                [
                    'name' => 'type',
                    'type' => 'hidden',
                    'value' => TransactionTypeEnum::INCOME->value,
                ],
                [
                    'name' => 'amount',
                    'type' => 'number',
                ],
            ],
        ]);

        CRUD::addField([
            'type'          => 'relationship',
            'name'          => 'expenses',
            'subfields'   => [
                [
                    'name' => 'label',
                    'type' => 'text',
                ],
                [
                    'name' => 'type',
                    'type' => 'hidden',
                    'value' => TransactionTypeEnum::EXPENSE->value,
                ],
                [
                    'name' => 'amount',
                    'type' => 'number',
                ],
            ],
        ]);
```

**(B) what you expected to happen**
Both incomes and expenses are stored in the database correctly

**(C) what happened**
When only incomes were entered and the expenses field was left empty. The incomes didn't get stored in the database.

**(D) what have you already tried to fix it?**
I found out the deletion happens in the `attachManyRelation` in the `Create` trait of the `CrudPanel`.
First the incomes are saved but after because the expenses field is empty the following code is triggered:
```
        if ($relationValues === null) {
            // the developer cleared the selection
            // we gonna clear all related values by setting up the value to the fallback id, to null or delete.
            $removedEntries = $modelInstance->where($relationForeignKey, $item->{$relationLocalKey});

            return $this->handleManyRelationItemRemoval($modelInstance, $removedEntries, $relationDetails, $relationForeignKey);
        }
```
The removed entries are solely defined by the foreign key, the type scope is neglected and all related records are deleted.

Could removed entries be altered to only clear the records of the relationship? I tested it like this
`$removedEntries = $relation;`
Or you could get the SQL from the relationship to execute directly on the database
`$relation->getQuery()->toSql()`

**Comment by pxpm** [comment]
Hey @madriegh thanks for the report.

I am sure I haven't take this use-case into consideration when I built the functionality. Sorry 🙃 

Will have a look at it and get back here with news. 

Cheers

**Comment by pxpm** [comment]
Hello @madriegh 

Thanks again, I think I was able to reproduce your scenario and indeed we had a problem there. 
One thing that I immediatelly thinked was using a polymorphic relation here, not sure but it seems it's the better fit here. 

Anyway, I saw the PR you sent, thanks for that too 🙏 
I've created a branch starting from it to include your commit and added a test for it too and some little fixes to make it easier for us to merge with confidence we are not breaking other stuff (🤞 🙃 )

I will be closing this and the PR, let's keep any further conversation on #4554 thread. 

Cheers
